### PR TITLE
Add node_modules cache in ci

### PR DESCRIPTION
**Pull request by wendevlin**
## Proposed change
- Cache node_modules in ci
  - key is bind by os, node version and yarn.lock hash
  - restore key is bind by os and node version

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by bramkragten** [comment]
You should use `actions/cache` and use keys to validate the cache

**Comment by bramkragten** [comment]
Make sure to also invalidate the cache completely if the node version or yarn version changes

**Comment by wendevlin** [comment]
> You should use `actions/cache` and use keys to validate the cache

Ok I'll try, it was just an expirment because yarn download always takes more than 1 minute. In `actions/setup-node` we define yarn as cache, but I am not sure if this is really caching anything.

**Comment by bramkragten** [comment]
> > You should use `actions/cache` and use keys to validate the cache
> 
> Ok I'll try, it was just an expirment because yarn download always takes more than 1 minute. In `actions/setup-node` we define yarn as cache, but I am not sure if this is really caching anything.

It does cache, but `node_modules` is not cached, as it is not recommended to cache `node_modules`.

https://github.com/actions/setup-node#caching-global-packages-data

**Review by silamon — DISMISSED** [review]

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -78,11 +92,19 @@ jobs:
...
+        id: setup-node
         uses: actions/setup-node@v4.1.0
         with:
           node-version-file: ".nvmrc"
           cache: yarn
+      - uses: actions/cache@v4
+        id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+        with:
+          path: "node_modules"
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}-${{ steps.setup-node.outputs.node-version }}
```
Comment (bramkragten) [inline .github/workflows/ci.yaml:104]:
```suggestion
          key: ${{ runner.os }}-yarn-${{ steps.setup-node.outputs.node-version }}-${{ hashFiles('**/yarn.lock') }}
```

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -60,11 +66,19 @@ jobs:
...
```
Comment (bramkragten) [inline .github/workflows/ci.yaml:78]:
```suggestion
          key: ${{ runner.os }}-yarn-${{ steps.setup-node.outputs.node-version }}-${{ hashFiles('**/yarn.lock') }}
```

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -26,14 +26,20 @@ jobs:
...
```
Comment (bramkragten) [inline .github/workflows/ci.yaml:38]:
```suggestion
          key: ${{ runner.os }}-yarn-${{ steps.setup-node.outputs.node-version }}-${{ hashFiles('**/yarn.lock') }}
```

**Review by bramkragten — DISMISSED** [review]

**Review by silamon — APPROVED** [review]

**Comment by steverep** [comment]
@wendevlin  and @bramkragten I'd argue this should be reverted as too dangerous and inefficient.  And there are better ways to address the problem.  Indulge my verbosity...

#### Reasons to revert
1. On inefficiency, you might have noticed now that we've got [2 caches for yarn](https://github.com/home-assistant/frontend/actions/caches?query=yarn): ~170 MB from the setup-node action and a new 240 MB one from this PR.  They are mostly storing the exact same things, except the latter is stored in zip archives in `~/.yarn/cache`, and thus faster to update and restore.
2. As noted above, it's not recommended to cache `node_modules`.  There's just too much that goes into it.  You've tried to mitigate that by adding the node version and yarn config file to the key, but there's easy to find holes there.  As examples, consider that updating one of our patch files in `.yarn/patches`, or switching to using corepack for yarn, both break that key and you get a hit where there shouldn't be.  Most importantly though, the output in `node_modules` is also the result of running build scripts for packages, and you don't know what they might do or what they depend upon.  For example, they might build a WASM or binary that depends on OS packages (which are also not part of the key).

#### How to reduce yarn install time
I also have noticed that it always takes ~1 minute in CI (and ~3 minutes locally for me after wiping `node_modules`).  However, if the packages are cached, almost none of that time is resolving and fetching, so the cache is doing its job.  The time is in the linking and building steps.  I looked into this a while ago and came up with 2 things to bring that time way down:
1. Get rid of dirty packages with ridiculous build steps.  By **far**, the culprit to focus on here is `app-datepicker`, which ends up including things like `nodemod`, `normalize-diacritics`, and `lit-ntml`, all of which spend time downloading more stuff we don't need.  Go ahead and run `rm -rf node_modules && yarn install --immutable --inline-builds` to see what I'm talking about.  I mentioned this in #14141, so if we prioritize switching to vaadin there, the bulk of that time gets eliminated.
2. Switch to yarn PNP (which virtualizes `node_modules`).  I tried it when I wrote the Webpack plugin for TLA, and I was impressed at how easy it was to get used to.  And it's also very fast (takes just a few seconds to install for that ~20 MB cache).  I believe I started a branch to implement it for HA, so I can revive and finish that if there's interest.  Some devs will get confused at first, but it's worth the switch IMO.  Biggest gotcha is that you cannot explore the dependencies without a compatible editor SDK, of which there's only a few.

Cheers 🥂

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -26,14 +26,20 @@ jobs:
...
+        id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+        with:
+          path: "node_modules"
+          key: ${{ runner.os }}-yarn-${{ hashFiles('.yarnrc.yml') }}-${{ steps.setup-node.outputs.node-version }}-${{ hashFiles('yarn.lock') }}
+          restore-keys: ${{ runner.os }}-yarn-${{ hashFiles('.yarnrc.yml') }}-${{ steps.setup-node.outputs.node-version }}
       - name: Install dependencies
+        if: steps.yarn-cache.outputs.cache-hit != 'true'
         run: yarn install --immutable
-      - name: Check for duplicate dependencies
-        run: yarn dedupe --check
```
Comment (steverep) [inline .github/workflows/ci.yaml:36]: Also, why remove this check?

**Comment by steverep** [comment]
FYI... just to show my work, I tested removing `app-datepicker` and I get a 76% reduction in clean install time (from 2m 41s to 38s).  GitHub runners are about 2x faster, so that'd probably mean under 20s in workflows just from that.

PNP would likely bring it down further to ~5s.

**Pull request merged**